tests/context.py

```diff
@@ -14,6 +14,13 @@
 
 import fitgrid
 
+# LMER fits differ slightly depending on the build/env. Cap allowable
+# variations like so
+#   np.allclose(actual, expected, atol=FIT_ATOL, rtol=FIT_RTOL)
+# atol comes into play for numbers < 1, rtol for numbers > 1
+FIT_ATOL = 0.001
```

Comment:
Should we make the absolute tolerance tighter, like `1e-6`? Ideally we'd set it to zero, but then the comparison would fail if `actual` or `expected` are zero.

Reply:
Yes that's better I was being lazy. I want to know two things from the tests: are the calculated values as expected, if not how far off are they? With a small atol==rtol==1e-3 test_utils_summary::test_summary throws two warnings and its a quick check that yes, something changed and yes the values are out of tolerance at 0.001 but not by much so the worst case is discrepancy is still not terribly serious in practice. With atol==rtol==1e-6  there are 65 warnings and I have to wade through them all to find where and what the biggest discerpancies are. So how about I tighten the tolerance as you suggest, and then the rewrite the test to accumulate and sort the discrepancies, before reporting. That will have the stricter test and still quickly answer the question about how far things are off.

Reply:
So now I have defined a two sets of atol, rtol values:

1. tolerable discrepancies, even very small ones, that merely trip warnings
2. intolerable discrepancies that raise an Exception and fail the pytest.

Strict warnings + looser fails is closer to Option 3 #162

For these values there are many small discrepancies between the two versions of r-lme4 and r-matrix we have run into so far but discrepancies large enough to warrant failing the pytest. 

```
# throw warnings on these 
FIT_ATOL = 1e-6 
FIT_RTOL = 1e-6

# fail on these
FIT_ATOL_FAIL = 1e-2
FIT_RTOL_FAIL = 1e-2
```



All warnings, if any are dumped. The fails, if any, are flagged X.

Here's what the output looks (**tolerances are temporarily set to throw both warnings and fails THESE ARE NOT REAL FAILS**)

Nevertheless, it is interesting that the discrepancies for the fails are not numerically large, they are small changes in small values but disproportionately large (`rtol`).

```
pytest -s tests/test_utils_summary.py::test_summarize
========================================================== test session starts ===========================================================
platform linux -- Python 3.6.9, pytest-5.3.2, py-1.8.0, pluggy-0.13.1
rootdir: /mnt/cube/home/turbach/TPU_Projects/fitgrid
collected 1 item                                                                                                                         

100% 2/2 [00:00<00:00, 21.82it/s]
100% 2/2 [00:00<00:00, 43.38it/s]
100% 2/2 [00:00<00:00, 41.52it/s]
100% 2/2 [00:00<00:00, 44.55it/s]
100% 2/2 [00:00<00:00, 41.30it/s]
100% 2/2 [00:00<00:00, 49.63it/s]
100% 2/2 [00:00<00:00, 40.52it/s]
100% 2/2 [00:00<00:00, 43.23it/s]
100% 2/2 [00:00<00:00, 44.74it/s]
100% 2/2 [00:00<00:00, 43.68it/s]
100% 2/2 [00:00<00:00, 44.59it/s]
100% 2/2 [00:00<00:00, 40.85it/s]
100% 2/2 [00:00<00:00, 36.96it/s]
100% 2/2 [00:00<00:00, 39.69it/s]
100% 2/2 [00:00<00:00, 39.29it/s]
100% 2/2 [00:00<00:00, 27.60it/s]
100% 2/2 [00:00<00:00, 26.91it/s]
100% 2/2 [00:00<00:00, 28.58it/s]
100% 2/2 [00:00<00:00, 26.11it/s]
100% 2/2 [00:00<00:00, 24.63it/s]
F

================================================================ FAILURES ================================================================
_____________________________________________________________ test_summarize _____________________________________________________________


<snip>

                    if oo_tol_fail.any():
                        fail_msg = (
                            f'Fitted values marked X are too far out of tolerance '
                            f'+/- ({FIT_ATOL} + {FIT_RTOL} * expected)'
                        )
>                       raise Exception(fail_msg)
E                       Exception: Fitted values marked X are too far out of tolerance +/- (0.0001 + 0.0001 * expected)

<snip>

tests/test_utils_summary.py::test_summarize
  /mnt/cube/home/turbach/TPU_Projects/fitgrid/tests/test_utils_summary.py:209: UserWarning: 
  ------------------------------------------------------------
  37 / 352 fitted vals out of tolerance: +/- (0.0001 + 0.0001 * expected)
                                                                                       actual     expected  abs_delta fail
  Time model                                       beta        key                                                        
  0    1 + continuous + (continuous | categorical) continuous  P-val    channel1     0.366931     0.367096   0.000165     
                                                   (Intercept) P-val    channel1     0.394986     0.395166   0.000181     
  1    1 + continuous + (continuous | categorical) (Intercept) P-val    channel0     0.237423     0.237606   0.000183     
       1 + (continuous | categorical)              (Intercept) P-val    channel0     0.982172     0.982423   0.000251     
  0    1 + continuous + (continuous | categorical) continuous  T-stat   channel1    -1.567500    -1.567191   0.000309     
  1    0 + continuous + (continuous | categorical) continuous  T-stat   channel0    -1.173861    -1.174178   0.000316     
  0    1 + continuous + (continuous | categorical) (Intercept) T-stat   channel1     1.429858     1.429518   0.000340     
       1 + (continuous | categorical)              (Intercept) SE       channel0     1.405398     1.405781   0.000383     
  1    1 + (continuous | categorical)              (Intercept) T-stat   channel0    -0.028242    -0.027845   0.000397     
  0    1 + continuous + (continuous | categorical) continuous  DF       channel1     0.970149     0.969564   0.000586     
                                                   (Intercept) DF       channel1     0.962947     0.962329   0.000618     
       1 + (continuous | categorical)              (Intercept) DF       channel0     2.963679     2.962921   0.000759     
  1    0 + continuous + (continuous | categorical) continuous  DF       channel0     1.104725     1.103561   0.001164     
       1 + continuous + (continuous | categorical) continuous  DF       channel0     3.255161     3.253966   0.001195     
  0    0 + continuous + (continuous | categorical) continuous  Estimate channel1    -7.479858    -7.481081   0.001223     
  1    1 + (continuous | categorical)              (Intercept) SE       channel0     8.934655     8.933294   0.001361     
  0    1 + (continuous | categorical)              (Intercept) T-stat   channel0    10.156822    10.154042   0.002780     
  1    1 + (continuous | categorical)              (Intercept) Estimate channel0    -0.252334    -0.248746   0.003589    X
       0 + continuous + (continuous | categorical) continuous  SE       channel0    12.744308    12.748530   0.004222     
       1 + continuous + (continuous | categorical) (Intercept) SE       channel0    24.710612    24.715215   0.004603     
                                                               DF       channel0     3.130235     3.125167   0.005069    X
  0    1 + (continuous | categorical)              (Intercept) sigma2   channel0    10.842424    10.848339   0.005915     
  1    1 + (continuous | categorical)              (Intercept) 2.5_ci   channel0   -17.763937   -17.757681   0.006256     
       1 + continuous + (continuous | categorical) (Intercept) 2.5_ci   channel0   -12.418315   -12.425208   0.006894     
       0 + continuous + (continuous | categorical) continuous  Estimate channel0   -14.960050   -14.969039   0.008989     
       1 + continuous + (continuous | categorical) (Intercept) 97.5_ci  channel0    84.445503    84.456653   0.011149     
  0    1 + continuous + (continuous | categorical) continuous  SE       channel1    65.528784    65.540584   0.011800     
                                                   (Intercept) SE       channel1    55.401020    55.413488   0.012468     
  1    0 + continuous + (continuous | categorical) continuous  2.5_ci   channel0   -39.938436   -39.955699   0.017263     
  0    1 + continuous + (continuous | categorical) (Intercept) 97.5_ci  channel1   187.799619   187.823030   0.023411     
                                                   continuous  97.5_ci  channel1    25.717657    25.742571   0.024914     
                                                   (Intercept) 2.5_ci   channel1   -29.368390   -29.393852   0.025462     
                                                   continuous  sigma2   channel1   310.601492   310.551110   0.050382     
                                                   (Intercept) sigma2   channel1   310.601492   310.551110   0.050382     
  1    0 + continuous + (continuous | categorical) continuous  sigma2   channel0   225.276469   225.197089   0.079379     
       1 + (continuous | categorical)              (Intercept) SSresid  channel0  1155.631600  1155.440753   0.190847     
       0 + continuous + (continuous | categorical) continuous  SSresid  channel0   950.882531   950.466408   0.416122     
  ------------------------------------------------------------

```

Reply:
@turbach This makes sense and is definitely a good compromise and a clear system. One thing that’s unclear to me though is how to react to these discrepancies. These tests tell us that the numerics change from version to version, but… what to do about it?

Reply:
Me neither. Documenting the possibility of small version discrepancies so users can monitor their own work is important. Alerting fitgrid maintainers to unexpected and potentially pathological changes via Travis fails is also important. Beyond that I don't know what to do. So far the small version discrepancies seem to me like benign skin growths ... unattractive but not medically significant.